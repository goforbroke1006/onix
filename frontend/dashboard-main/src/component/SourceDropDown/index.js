import React from "react";
import PropTypes from "prop-types";
import {Select} from "antd";

const {Option} = Select;

export default class SourceDropDown extends React.Component {
    static propTypes = {
        onChange: PropTypes.func,
        provider: PropTypes.object,
    };

    constructor(props) {
        super(props);
        this.state = {
            items: [],
            value: "",
        };
    }

    componentDidMount() {
        this.loadSources();
    }

    onChange = (sourceId) => {
        sourceId = parseInt(sourceId);

        this.setState({value: sourceId});

        if (this.props.onChange)
            this.props.onChange(sourceId);
    }

    render() {
        let emptyOption;
        if (this.state.items.length === 0) {
            emptyOption = (<Option value={""}>-- no data --</Option>)
        } else {
            emptyOption = (<Option value={""}>-- none --</Option>)
        }

        return (
            <Select defaultValue={this.state.value} style={{width: 600}} onChange={this.onChange}>
                {emptyOption}
                {this.state.items.map((object, index) => {
                    return (
                        <Option key={`source-${index}`}
                                value={object.id}>{`${object.title} [${object.kind}] ${object.address}`}</Option>
                    )
                })}
            </Select>
        )
    }

    loadSources = () => {
        return this.props.provider.loadSourcesList()
            .then(data => {
                this.setState({items: data});
            })
            .catch((err) => {
                console.warn(err);
                this.setState({items: []});
            })
    }
}
