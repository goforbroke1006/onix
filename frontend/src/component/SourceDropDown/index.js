import React from "react";

class SourceDropDown extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            items: [],
            value: "",
        };
    }

    componentDidMount() {
        this.loadSources();
    }

    onChange = (event) => {
        let sourceId = event.target.value;

        this.setState({value: sourceId});

        if (this.props.onChange)
            this.props.onChange(sourceId);
    }

    render() {
        return (
            <select onChange={this.onChange} value={this.state.value}>
                <option>----</option>
                {this.state.items.map((object, index) => {
                    return (
                        <option key={"source-" + index} value={object.id}>
                            {object.title} [{object.kind}] {object.address}
                        </option>
                    )
                })}
            </select>
        )
    }

    loadSources = () => {
        let baseUrl = process.env.REACT_APP_API_DASHBOARD_MAIN_BASE_ADDR ?? 'http://127.0.0.1:8082/api/dashboard-main';
        fetch(baseUrl + "/source")
            .then(response => response.json())
            .then(data => this.setState({items: data}))
    }
}

export default SourceDropDown;