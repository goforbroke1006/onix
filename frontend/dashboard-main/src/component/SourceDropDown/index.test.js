import React from "react";
import {mount, shallow} from "enzyme";

import SourceDropDown from "./index";

describe('<SourceDropDown /> rendering', () => {
    it('renders correctly', () => {
        const client = new MockDashboardMainApiClient();
        const wrapper = mount(<SourceDropDown provider={client}/>);

        fakeSourcesListPromise.then(() => {
            wrapper.update();
            expect(wrapper).toMatchSnapshot();
        });
    });

    it('onChange callback works', () => {
        const expectedValue = 101;
        let actualValue = 0;
        let onChangeSpy = function (sourceId) {
            actualValue = sourceId;
        }

        const wrapper = shallow(<SourceDropDown provider={new MockDashboardMainApiClient()} onChange={onChangeSpy}/>);
        wrapper.find('select').at(0).simulate('change', {
            target: {value: '101', name: 'item1'}
        });

        expect(actualValue).toEqual(expectedValue);
    });
});


const fakeSourcesListPromise = new Promise((resolve, reject) => {
    const fakeSourcesList = [
        {id: 101, title: "Source 1", kind: "prometheus", address: "http://foo-bar:4001/"},
        {id: 102, title: "Source 2", kind: "influxdb", address: "http://foo-bar:4002/"},
        {id: 103, title: "Source 3", kind: "prometheus", address: "http://foo-bar:4003/"},
    ];

    resolve(fakeSourcesList)
});

class MockDashboardMainApiClient {
    loadSourcesList() {
        return fakeSourcesListPromise;
    }
}