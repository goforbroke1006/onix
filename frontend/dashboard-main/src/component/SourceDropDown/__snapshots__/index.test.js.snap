// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SourceDropDown /> rendering renders correctly 1`] = `
<ForwardRef(InternalSelect)
  defaultValue=""
  onChange={[Function]}
  style={
    Object {
      "width": 600,
    }
  }
>
  <Option
    value=""
  >
    -- none --
  </Option>
  <Option
    key="source-0"
    value={101}
  >
    Source 1 [prometheus] http://foo-bar:4001/
  </Option>
  <Option
    key="source-1"
    value={102}
  >
    Source 2 [influxdb] http://foo-bar:4002/
  </Option>
  <Option
    key="source-2"
    value={103}
  >
    Source 3 [prometheus] http://foo-bar:4003/
  </Option>
</ForwardRef(InternalSelect)>
`;
